### ISSUE TITLE
Thread Safe Planner

### ISSUE THREAD

**Issue by x42** [FFTW/fftw3#16]
http://www.fftw.org/fftw3_doc/Thread-safety.html mentions 
 > We do not think this should be an important restriction

I hope I can convince you otherwise.

Problem Description
-------------------------

fftw3 is used by a variety of audio plugins.

Those plugins are loaded into the host's memory-space (usually an audio workstation). The host has limited control of what the plugin does internally, and the plugins do not know about each other.

There is no way to ensure that two independent plugins which are linked against libfftw do not run the shared planner simultaneously. Nor is there a possibility to control this on host application level.

When two independent plugins create fftw plans the application usually segfaults or similar undesired effects manifest.

Possible solutions for this include:

1. Statically link plugins against libfftw. Every plugin will have its own copy. The plans are not shared with other plugins (which is mostly fine). This still requires a bit of special attention: (fftw symbol visibility needs to be overridden for static links and the plugin must protect its planning routings for multiple instances of itself). Furthermore distributors must honor that (special built of fftw + static link). -- It is very unlikely that both plugin-authors and various gnu/linux-distributors do get this right (most distros dislike static linking) for the growing number of audio-plugins using fftw.

2. process separate all plugins in the host. That is not a viable option for Digital Audio Workstations where low-latency is important, context switches (particularly realtime thread) heavy and inter-process communication does not scale (compared to shared memory), especially so if the DAW does not limit audio track or channel count.

3. Discourage use of fftw for audio-plugins or even refuse to load plugis using it in the host. -- not the best idea :)

4. Ship a special (ABI compatible) build of libfftw with the host application which protects the planner. Plugins in the same memory space will use the already loaded library. This requires patching libfftw, but when doing so... why not do it upstream directly. Otherwise it has similar issues as (1).

Discussion
--------------

The issue at hand is not limited to audio-application, there are likely other applications with similar problems out there (gnu-octave comes to mind, but I don't know for certain).

As the Thread-safety page mentions, it's as simple as 
> wrap a semaphore lock around any calls to the planner

Is there some good reason why libfftw does not do this by default?

Existing applications should not be affected by this (they're not supposed to call the planner from different threads), but that change would make all the difference for multi-threaded plugin hosts.

I suppose it could be a bit of work to wrap all planner entry-points with a semaphore, yet there may be a neat simple solution using `#define`.

I'll be happy to look into this, but before going that way, I'd like to ask if such a change would be accepted by fftw or if there is an even better solution planned for future version that will make fftw's planner thread-safe.

yours truly,
robin - for the linux-audio community and for himself

---

Notable audio plugins using fftw3:
 http://calf.sourceforge.net/
 http://factorial.hu/plugins/lv2/ir
 http://guitarix.sourceforge.net/
 http://breakfastquay.com/rubberband/
 http://plugin.org.uk/
 http://zynaddsubfx.sourceforge.net/
 https://github.com/x42/meters.lv2
...

Notable affected plugin hosts:
 http://ardour.org/
 http://qtractor.sourceforge.net/
 https://github.com/falkTX/Carla/
...

see also https://community.ardour.org/node/8271

**Comment by jrigg** [comment]
Has there been any more progress on this?

**Comment by harryhaaren** [comment]
Hi All,

A ping on this issue - this issue is discussed periodically on IRC in various linux-audio related channels, and would be really nice to have fixed.

I'm currently looking towards using a different library for FFT for an upcoming project due to this issue possibly causing crashes in software designed for on-stage live performance.

The solution proposed with both callbacks set in one function looks a good fix to me.
Cheers, -Harry

**Comment by magnetophon** [comment]
I'll add my voice to the choir!  :)

**Comment by rdolbeau** [comment]
WRT a simple global lock - there's no need for pthread or anything heavy. A simple by-the-book lock should be enough (... I think). GCC defines intrinsics for atomic operation (and they are available on most architectures), so it's quite trivial to implement a ticket lock on a pair of statically-initialized to 0 integer. I've pushed a trivial example on <https://github.com/rdolbeau/fftw3/tree/mutualexclusion>. Just add "--enable-mutualexclusionplan" to configure, and the planner _should_ be "thread safe" (as in, serialized :-). Beware: active wait, and not really tested at all.

**Comment by matteo-frigo** [comment]
Current status on this issue:

The next fftw-3.3.5 will support fftw_set_planner_hooks(before, after, arg).
Before planning, fftw calls before(arg), and after planning it calls after(arg).

fftw_set_planner_hooks() is not thread safe; the user is responsible to ensure that it is called once.
@rdolbeau : it is a bad idea to add a gcc dependency (__sync* intrinsics) so deep in FFTW---whatever the API we end up with, we must support it everywhere (including e.g. Windows without pthreads).

Please offer feedback on whether or not this solves the problem.  If it does I'll document it in the manual, and if it does not I will remove the hook completely.

**Comment by rdolbeau** [comment]
@matteo-frigo The __sync* intrinsics are supported by ICC and CLANG as well. For windows, there is a similar mechanism (InterlockedAdd ?). Basically, you only need a way to do "x = *y; (*y)++" atomically to implement a ticket lock. And it doesn't require any library support beyond the compiler itself, so it avoids a dependency on pthread.
Hooks can work, but require application-level support.

**Comment by x42** [comment]
A hook for locking offers a nice flexible API, but that requires each and every application to explicitly use  it.  Also in the Host+Plugins case it can quickly end up being a mess: Who is responsible to register the hooks? (e.g. A host is not using fftw, but some plugins loaded into the host do. Plugins can be un/reloaded.)

I'm all for hooks, but they should by default point to an internal simple lock (a ticket spinlock is fine. I like Romain's implementation.)

The reasoning is as follows:

*   Old applications won't call the planner in parallel (it was never thread safe). The lock-wait won't ever be used and there won't be a difference (only a handful of CPU cycles bumping the ticket's integer)
*   fftw is thread safe by default.
*  advanced applications can use the hook to their advantage: register custom functions to call back "planning is done" or disable the lock for whatever reason or add a different (non spinning) implementation...

**Comment by matteo-frigo** [comment]
The spinlock implementation is a bad idea for at least two reasons.

First, the planner may take minutes, and the last thing you want to do is spin while the planner is running.

Second, we aim to be much more portable than "icc" and "clang".  Look at kernel/cycle.h for a partial list of systems where the lock hack is expected to work.

The current status is that the fftw core (e.g., non-simd) requires ANSI-C and nothing else (no threads, no c99, no nothing), and the FFTW API works in the same way in all systems.  It will take an extraordinary amount of evidence to convince us to break this property.

**Comment by matteo-frigo** [comment]
How about this.

To the threaded api, we add a call fftw_make_planner_thread_safe() (or something) that atomically installs the appropriate hooks, and is guaranteed to be atomic and idempotent.

Plugins that care about thread safety will have to call fftw_make_planner_thread_safe().  The fftw core does not depend upon pthreads.

If we go down this path, I will remove the hooks for lack of a use case.

**Comment by davisking** [comment]
Why is there global state in FFTW at all?  By that I mean, why doesn't the interaction with FFTW begin by allocating some kind of context structure which is passed to all the other FFTW functions and contains the "global" state.  This would avoid any need for behind the scenes locks.

**Comment by matteo-frigo** [comment]
The problem is not to argue what made or didn't make sense in 1997 when
fftw started.  The problem is to decide how to move forward productively
from where we are.

**Comment by davisking** [comment]
Obviously don't break the existing API.  But, for example, why not add a version of fftw_plan_dft_1d() that doesn't have any global state?   Call it fftw_plan_dft_1d_r() and make it either take an additional context object.  Or make it simply allocate and deallocate whatever it needs internally.

**Comment by x42** [comment]
`fftw_make_planner_thread_safe()` is so far the best option on the table.
It still requires application level support but that may be an acceptable compromise.
 
Why can't this be done by default if fftw3 is compiled with --enable-threads?

**Comment by x42** [comment]
@davisking  `fftw_plan_dft_1d`  and `fftwf_plan_r2r_1d` and `fftwf_plan_dft_r2c_1d` and probably a few more... special casing a single plan is not a solution.

A similar workaround is just copy all of fftw into your project (the global state won't be shared with other projects that use system-wide libfftw.so, you only have to worry about multiple instances of your own project which is trivial given that you have control over the fftw configuration as part of your project). Then again almost all gnu/linux distros refuse to package a project like that and instead ask to get things upstream..

**Comment by davisking** [comment]
Yes, I mean do it for all the planners.  Everything else we are talking about are a bunch of hacks to work around the problem imposed by a lack of reentrant planning functions in the API.  As you and others have pointed out, the other solution involves encouraging fftw users to do things they will either refuse to do or will complicate their build systems and likely lead to package managers (or other downstream users) accidentally building things without the appropriate thread safety switches turned on.

Is it really that hard to push the global variables used by the planners into an argument?  I haven't looked at FFTW's code but I can't imagine it's a huge deal to refactor it into that form.  Then you implement the existing non-thread-safe APIs in terms of the reentrant APIs to avoid breaking backwards compatibility.

**Comment by x42** [comment]
@matteo-frigo  Does the problem also exist if fftw is built without `--enable-threads` but an application calls fftw_plan_*  from different threads?

**Comment by matteo-frigo** [comment]
@x42 (et al.):

Whether this is a "problem" or not depends on your perspective.

FFTW is meant primarily to be called from one thread and use threads internally to speed up the computation.  This is a common mode of operation of numerical libraries: you write your sequential program that calls fftw/atlas/whatever, and the numerical library handles the parallelism.  The core of FFTW consists of one library that does not depend on pthreads (e.g., because the user may want to use openmp instead of pthreads, or not want threads at all).  --enable-threads produces *another* library libfftw3_threads that exports a function fftw_init_threads() that installs the threaded solvers into the fftw core.  Such a library is optional; --enable-openmp installs openmp solvers instead (and other awful variants exists; for example support for the Cell processor adds solvers that use the coprocessors).

Thus, the intended usage of fftw is the opposite of what the commenters on this ticket have in mind,  that is, calling fftw from multiple threads and run fftw independently in each thread.

The root of the problem is that it is hard to have both setups.  Since fftw --enable-threads uses threads internally, the internal threads are a shared resource that is hard to manage if we allow multiple threads to enter the planner at the same time.  To put it another way, threads do not compose.  Attempts to provide a better abstraction parallelism (such as Cilk or TBB) have not won in the marketplace, so we are stuck with the problems of threads.

I think that the suggestion of passing a planner "object" to all planning function is fundamentally misguided because it attacks the syntax of the problem but not its essence.  At its root, threaded fftw has a bunch of global resources (internal threads) that cannot be abstracted away via syntactic sugar.  Thus, this suggestion is fundamentally incompatible with --enable-threads. (For the record, this suggestion would not be hard to implement. fftw internally does in fact pass an explicit planner "object" pointer everywhere in the way suggested by davisking.  The API layer deliberately uses a global planner precisely not to sweep the real problem under the rug.)

I think that my proposal of fftw_make_planner_thread_safe() is the only practical workaround: it recognizes that threads do not compose, that fftw is setup so that parallelism is inside fftw and not outside, and it works around the lack of composability via a lock, and it does not break any existing code. This proposal is currently implemented in git, so folks may want to try it out and see if it works for them.

**Comment by x42** [comment]
Nice, I can confirm that `fftwf_make_planner_thread_safe()` works and resolves the issue.

However, so far there is no C-header for that function, and ` -lfftw3f_threads` need to be added manually. It'd be great if `pkg-config  --libs fftw3f` could take care of that.

The test was performed on debian/stable, fftw3 git 
```
# git describe --tags
fftw-3.3.4-55-g8cd9bfa
# ./configure --enable-single --enable-sse --enable-avx --enable-shared --enable-static --enable-threads
```

PS. How can one easily check if that function is available at compile time? Is that best done during configuration `pkg-config --atleast-version=3.3.5 fftw3f` or will there be a header define in FFTW3 itself?

**Comment by x42** [comment]
PS. the test I ran is https://gist.github.com/x42/07391de61b99e872ad9d
without `fftwf_make_planner_thread_safe()` there are regular crashes.

**Comment by x42** [comment]
Any news on this?

looks like all that's left to do is adding `fftwf_make_planner_thread_safe` to the header and  `-lfftw3f_threads` to the pkg-config libs.

**Comment by Parabola25** [comment]
Hi guys I m no programmer and got very few basics skills with linux but I love Ardour and calf plugins. Could someone help me resolve the crash when loading some .ardour saved session? I dont understand how this fftw work..! Thx

**Comment by matteo-frigo** [comment]
What do you mean by "adding to the header"?
fftwf_make_planner_thread_safe() is already in the header fftw3.h:

   FFTW_EXTERN void X(make_planner_thread_safe)(void);

**Comment by x42** [comment]
indeed, there it is. sorry for the noise. The include works now. `-lfftw3f_threads` is still needed though.
Even when configured with `--enable-threads` the .pc file only references `-lfftw3` (or `-lfftw3f` in single mode).

PS. I suppose last time I tested the system-wide `fftw3.h` was preferred over the the local version (using PKG_CONFIG_PATH).

**Comment by matteo-frigo** [comment]
We cannot add -lfftw*_threads to the pkg-config file, because this would
force all applications to link to the pthreads library.  This whole topic
was about how to avoid forcing such a dependency.

**Comment by x42** [comment]
How do you envisage to solve this?  A second pkg-config file `fftw3[f]-threaded`?

**Comment by x42** [comment]
I'll close this. The original issue is fixed in fftw3-git.

It'd be nice if there'd be an easy way to find out if a given version of fftw in a system has fftw3_threads.
Best solution I can currently think of is a 2nd `.pc`  pkg-config file, but that's a different story.

Thanks to all involved involved in fixing this issue.

**Comment by stevengj** [comment]
Doesn't `fftw_destroy_plan` also need a lock, since it accesses shared state for trig tables?

**Comment by matteo-frigo** [comment]
@stevengj Yes, it's a bug, I'll fix it

**Comment by rlcamp** [comment]
I think there is still room for improvement here. Perhaps an FFTW_NO_GLOBAL_STATE planning flag? I'm willing to wait a hair longer for an FFTW_ESTIMATE if it means planning can happen indiscriminately in independently-developed plugins without counting on libfftw3f_threads being available.

Also: in light of the fftw_destroy_plan bug fixed in 3.3.6, and the fact that 3.3.5 is still the newest version in some major repositories: is there a way to check, at compile time, whether 3.3.5 or 3.3.6 is being linked against? There is the fftwf_version macro, but it's a string, you can't test an inequality against it in the C preprocessor.
